IABTechLab/uid2-self-serve-portal
# User/Participant/Role relationship backend

- Migration to create `userRoles` table
- Migration to add column to `usersToParticipantRoles` table, using default value of 1 (Admin)
- New `UserRole` and `UserToParticipantRole` entities
- Relate the `UserToParticipantRole` entity to `UserRole`, `User` and `Participant`.
- Alter participant/user creation methods so that the three-way mapping is also updated. Objection doesn't seem to handle this three-way mapping very well, so we've opted to manually insert into the mapping table rather than using graph inserts.
- Simplify some parsers to only parse the actual request from the front-end, rather than the request + properties we add in the back-end.
- Some minor refactoring out of router file
- Some minor renaming for clarity
- Modify `createUser` test method to accept a participant/role mapping, rather than just a `participantId`.

## Manual testing
1. Run the migrations to get the latest schema. 
2. Observe that all rows in `usersToParticipantRoles` have a `userRoleId` of 1.
3. Add yourself to another participant manually into the db, optionally with a different `userRoleId`. For me, this was done via:
    ```
    insert into dbo.userstoParticipantRoles (userId, participantId, userRoleId) values (3, 9, 3)
    ```
4. Log into the UI and check the response from `http://localhost:3000/api/users/current`. It should include the `currentUserRoleIds` array (as well as a separate `Participants` array), e.g.:
    ```
    "id": 3,
    "email": "abby@example.com",
    ...
    "participantId": 8, // will be changed/removed in future
    "participants": [
        {
            "id": 8,
            "name": "Participant A",
            ...
            "currentUserRoleIds": [
                1
            ]
        },
        {
            "id": 9,
            "name": "Participant B",
            ...
            "currentUserRoleIds": [
                3
            ]
        }
    ],
    ...
    ```
